File: src/main/java/komu/blunt/core/PatternPath.java
package komu.blunt.core;

import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.List;

public final class PatternPath {
    
    public static final PatternPath EMPTY = new PatternPath(null, -1);
    private final PatternPath parent;
    private final int index;
    
    private PatternPath(PatternPath parent, int index) {
        this.parent = parent;
        this.index = index;
    }

    public PatternPath extend(int index) {
        return new PatternPath(this, index);
    }
    
    public List<Integer> indices() {
        List<Integer> indices = new ArrayList<>();
        
        for (PatternPath p = this; p.parent != null; p = p.parent) 
            indices.add(p.index);

        return Lists.reverse(indices);
    }

    @Override
    public String toString() {
        return indices().toString();
    }
}

File: src/main/kotlin/komu/blunt/core/PatternPath.kt
package komu.blunt.core

import java.util.*

class PatternPath private constructor(private val parent: PatternPath?, private val index: Int) {

    companion object {
        val EMPTY = PatternPath(null, -1)
    }

    fun extend(index: Int) = PatternPath(this, index)

    val indices: List<Int>
        get() {
            val indices = ArrayList<Int>()

            var p = this
            while (true) {
                val parent = p.parent ?: break
                indices.add(p.index)
                p = parent
            }

            indices.reverse()
            return indices
        }

    override fun toString() = indices.toString()
}

